### ISSUE TITLE
Give the ability to know if a path does not exists

### ISSUE THREAD

**Issue by jonajames** [wwkimball/yamlpath#211]
**Is your feature request related to a problem? Please describe.**
When given a path that does not exists `Processor.get_nodes()` either:

1. returns `None` if given the parameter `mustexists = False`
2. raises `YAMLPathException` if `mustexists = True`

Option **1** does not give the ability to know if the path is nonexistent or if it's just `None`, wich is a perfectly valid value in YAML. 
At the same time, using option **2**, catching the exception may hide the fact that the path given contains some errors.

**Describe the solution you'd like**
A simple and non-breakng solution is to sublclass `YAMLPathException` and raise different subclasses based on context.

```python
class UnmatchedYAMLPathException(YAMLPathException):
  pass
```

**Describe alternatives you've considered**
The current workarounds are either setting the `default_value` to something that can never be in YAML, or checking the description of the exception. Both solution are hacky.

I can provide a PR if needed

**Comment by wwkimball** [comment]
Calling something "hacky" is opinionated.  It's quite common to parse error message text to take different actions or better tailor the message text for end-users versus developers.  Even this project does it to create more useful errors that bubble up from ruamel.yaml or Python.

That said, I'm always happy when someone takes their time to provide feedback and I look at every opportunity to make this project even better!  I like the idea of offering more specialized exception classes.  You're right; I may lean on YAMLPathException too heavily purely out of habit/convenience.

However, there's a material flaw in your case that we need to unpack before we start changing any code.  Unless you've hit on an edge-case which you aren't sharing, your description of the behavior of `mustexist` is backwards.  Here's proof:

### 211.yaml
```yaml
---
this:
  path:
    exists: in the YAML file
```

### 211.py (largely copied from [the documentation](https://github.com/wwkimball/yamlpath/wiki/Python-Library))
```python
from types import SimpleNamespace

from yamlpath.common import Parsers
from yamlpath.wrappers import ConsolePrinter
from yamlpath.exceptions import YAMLPathException
from yamlpath import Processor, YAMLPath


# The various classes of this library must be able to write messages somewhere
# when things go bad.  This project provides a CLI-centric logging class named
# ConsolePrinter.  Even when not writing a CLI tool, you must still configure
# and pass ConsolePrinter or a class of your own with the same public API.  For
# just muting logging output -- except for unrecoverable errors -- you can use
# this simple configuration object:
logging_args = SimpleNamespace(quiet=False, verbose=False, debug=False)
log = ConsolePrinter(logging_args)

# Prep the YAML parser and round-trip editor (tweak to your needs).  You do not
# have to use Parsers.get_yaml_editor() but you must create a properly-
# configured instance of ruamel.yaml.YAML.
yaml = Parsers.get_yaml_editor()

# At this point, you'd load or parse your YAML file, stream, or string.  This
# example demonstrates loading YAML data from an external file.  You could also
# use the same function to load data from STDIN or even a String variable.  See
# the Parser class for more detail.
yaml_file = "211.yaml"
(yaml_data, doc_loaded) = Parsers.get_yaml_data(yaml, log, yaml_file)
if not doc_loaded:
    # There was an issue loading the file; an error message has already been
    # printed via ConsolePrinter.
    exit(1)

# Pass the logging facility and parsed YAML data to the YAMLPath Processor
processor = Processor(log, yaml_data)

# At this point, the Processor is ready to handle YAML Paths

queries = (
	(YAMLPath("this.path.exists"), True),
	(YAMLPath("this.path.exists"), False),
	(YAMLPath("this.path.does.not.exist"), True),
	(YAMLPath("this.path.does.not.exist"), False),
)
for yaml_path, must_exist in queries:
	try:
		for node_coordinate in processor.get_nodes(yaml_path, mustexist=must_exist, default_value='not found'):
			data = node_coordinate.unwrapped_node
			log.info(f"Got '{data}' from '{yaml_path}' when mustexist={must_exist}.")
			# Do something with each node_coordinate.node (the actual data)
	except YAMLPathException as ex:
		# If merely retrieving data, this exception may be deemed non-critical
		# unless your later code absolutely depends upon a result.
		log.error(f"Got error seeking {yaml_path} when mustexist={must_exist}:  {ex}")
```

### Output
```
Got 'in the YAML file' from 'this.path.exists' when mustexist=True.
Got 'in the YAML file' from 'this.path.exists' when mustexist=False.
ERROR:  Got error seeking this.path.does.not.exist when mustexist=True:  Required YAML Path does not match any nodes, 'this.path.does.not.exist'.
Please try --help for more information.
Got 'not found' from 'this.path.does.not.exist' when mustexist=False.
```

As you can see, when `mustexist=True` and the path does *not* exist, YAMLPathException is raised.  This is the expected behavior because the YAML Path points to data which does not exist.  On the other hand, when `mustexist=False`, no exception is raised and the `default_value` is returned, which is `None` when otherwise not provided and [which is detailed in the documentation](https://github.com/wwkimball/yamlpath/wiki/processor.py).

If you can demonstrate that you've found a way to get `mustexist` to operate backwards, I'm very interested in how you've done it.  If there's a bug in my code, I very much want to fix it.

**Comment by jonajames** [comment]
Sorry, I simply swapped the behaviour of the `mustexists` parameter in the description (now edited).

While the solution I'm using right now is in fact to catch the exception and check the corresponding message, here are some reasons why I don't think that's a good practice:

- error messages are not documented, so one must retrieve them by searching in the source code
- as they are not part of the interface they may be subject to changes anytime
- different messages are used in different functions although they have the same meaning 
  - https://github.com/wwkimball/yamlpath/blob/master/yamlpath/processor.py#L109
  - https://github.com/wwkimball/yamlpath/blob/master/yamlpath/processor.py#L177
- `StopIteration` is already the common way for python generators to indicate when they are empty

This is what I've ended up doing:
```python
try:
    for node in processor.get_nodes(path, mustexist=True):
        yield node
except YAMLPathException as yaml_exception:
    if yaml_exception.user_message == "Required YAML Path does not match any nodes":
        raise StopIteration()
    raise yaml_exception
```

**Comment by jonajames** [comment]
Replying to your challenge:
I'am not trying to read a specific file, but to read all files in a folder (like with '*').
If one file exists, one file is returned, one file will be opened; two files, two returned, two opened.
Why should zero be different? I don't want my application to crash when given empty folders.
The function is called `get_nodes` so I expect it to give out nodes, not caring about how many there are.
If I go searching in a database I expect rows to be returned, I don't get SQLExceptions for empty tables...
Also with files I tipically have the ability to check if a file exists or not before trying to open it and 
when the opening of a file fails it may be because the file does not exists, because of insufficient permissions and different errors are returned in different situations.

**Comment by wwkimball** [comment]
Let's take a look at what happens when file operations are performed using wildcards against directories which contain no files:

```
$ ls -la
total 8.0K
drwxr-xr-x  2 william william 4.0K Jan 26 13:20 ./
drwxr-xr-x 21 william william 4.0K Jan 26 13:20 ../

$ cat *
cat: '*': No such file or directory

$ echo $?
1
```

The case of zero files is indeed a special case.  As we can see, this (and other) commands emit an error akin to "File not found" and also abend with an error exit state (`1`).  Even using a wildcard (`*`), with zero files, the error condition is met.  This is the normal, expected behavior of any command which attempts to access a resource by path, that does not exist.

When a user tells `get_nodes` that the supplied YAML Path must exist in the data -- by setting `mustexist=True` -- they are entering into a contract; the supplied path must yield at least one node.  If it does not, the code necessarily deems the supplied YAML Path to be invalid.  As such, it emits an informative exception.

That said, the example of performing SQL queries against empty tables is more interesting.  As long as the syntax is valid, a query against an empty table will not generate an error; the user will instead get an empty record set.  Further, the idea of testing a file-system for the presence of a file and permissions to it before attempting to open a file is indeed best-practice pattern.

I still can't get to `StopIteration` because again, the method cannot return a generator -- not even an empty one -- when the requested data structure simply does not exist.  However, let's consider alternatives.  Some possibilities include:

1. Raise a new custom exception like you'd originally proposed when `get_nodes` and `set_value` must abort.  I like your suggestion of raising `UnmatchedYAMLPathException` or similar.
2. Provide a new method on `Processor` like `exists` which takes a YAML Path and simply indicates whether or not the indicated node(s) exists in the data.
3. Provide a new named argument to `get_nodes` (and `set_value`) which allows for an alternative handling of the "not found" case.  Something like `softfail=True` (defaults to `False`).  When `True`, no exception would be raised in this case; there would simply be no yielded generator.
4. Do nothing; you have already solved this case by intercepting the exception message and translating the condition to fit your specific use-case.

Any combination of these would naturally be an option, as well.  How would you like to see this resolved?

**Comment by jonajames** [comment]
I don't think that shell glob patterns are a good example here. That's the bash behaviour expanding globs to hemselves if ther's no match. And cat raises the error. find does not display this behaviour. But I don't want/need to convince you.
I get that that's the expected behaviour of `mustexist`, but there's not really another option: not setting that value is problematic in it's ways (it returns values that don't exist in the structure and changes the original structure).

Solution 1 is probably the simplest to implement and it doesn't break any existing behaviour.

**Comment by wwkimball** [comment]
This new feature has been added.  Please try out release version 3.8.0.  It not only adds 6 new Exceptions, but the Processor now has a new `exists` method which will indicate whether a given YAML Path would resolve to at least one node.  Thanks!!
